contracts/contracts/strategies/BaseCurveStrategy.sol

```diff
@@ -288,20 +359,135 @@ abstract contract BaseCurveStrategy is InitializableAbstractStrategy {
+        // Approve the Curve pool, eg 3Pool, to transfer an asset (required for adding liquidity)
+        asset.safeApprove(CURVE_POOL, 0);
+        asset.safeApprove(CURVE_POOL, type(uint256).max);
     }
 
     function _approveBase() internal virtual;
 
+    /***************************************
+                Curve pool helpers
+    ****************************************/
+
+    /**
+     * @dev Get the Curve pool index of the asset.
+     * This is reading from immutable variables to avoid costly storage reads.
+     */
+    function _getCoinIndex(address _asset)
+        internal
+        view
+        returns (uint256 coinIndex)
+    {
+        require(_asset != address(0), "Invalid asset");
+        if (_asset == coin0) {
+            return 0;
+        } else if (_asset == coin1) {
+            return 1;
+        } else if (_asset == coin2) {
+            return 2;
+        }
+        revert("Unsupported asset");
+    }
+
+    /**
+     * @dev Get the number of decimals of the asset token.
+     * This is reading from immutable variables to avoid costly storage reads.
+     */
+    function _getAssetDecimals(address _asset)
+        internal
+        view
+        returns (uint256 decimals)
+    {
+        // This check is needed for Curve pools with only two assets as
+        // coin2, the third coin, will be address(0)
+        require(_asset != address(0), "Invalid asset");
+
+        if (_asset == coin0) {
+            return decimals0;
+        } else if (_asset == coin1) {
+            return decimals1;
+        } else if (_asset == coin2) {
+            return decimals2;
+        }
+        revert("Unsupported asset");
+    }
+
     /**
-     * @dev Get the index of the coin
+     * @dev Get the asset token address for a given Curve pool index.
+     * @param _coinIndex Curve pool index
+     * This is reading from immutable variables to avoid costly storage reads.
      */
-    function _getCoinIndex(address _asset) internal view returns (uint256) {
-        for (uint256 i = 0; i < 3; i++) {
-            if (assetsMapped[i] == _asset) return i;
+    function _getAsset(uint256 _coinIndex)
+        internal
+        view
+        returns (address asset)
+    {
+        if (_coinIndex == 0) {
+            return coin0;
+        } else if (_coinIndex == 1) {
+            return coin1;
+        } else if (_coinIndex == 2) {
+            return coin2;
         }
-        revert("Invalid 3pool asset");
+        revert("Invalid coin index");
+    }
+
+    /***************************************
+            Asset/coin validation
+    ****************************************/
+
+    /**
+     * @notice Retuns bool indicating whether vault asset is supported by the strategy
+     * @param _vaultAsset Address of the vault asset
+     */
+    function supportsAsset(address _vaultAsset)
+        external
+        view
+        override
+        returns (bool result)
+    {
+        result = _curveSupportedCoin(_vaultAsset);
```

Comment:
You could do smth like: 
```
try getAssetDecimals(_vaultAsset) returns (uint256 v) {
   result = true;
} catch (bytes memory /*lowLevelData*/) {
   result = false;
}
```

and cancel the `_curveSupportedCoin(address _coin)` function. Though not sure how much gas that consumes it might require something similar to [Balancer's static call](https://github.com/balancer/balancer-v2-monorepo/blob/9c1574b7d8c4f5040a016cdf79b9d2cc47364fd9/pkg/pool-utils/contracts/lib/VaultReentrancyLib.sol#L78-L80) that limits the gas consumption.

Another thing is that `_curveSupportedCoin` insinuates a coin is supported by Curve pool, while `supportsAsset` suggests the coin is supported by the asset. I know those things mean the same value across all assets for now, but that could change, when/if we support only a subset of Curve pool's assets.